### PR TITLE
Fix: bug of out_band when kpar > 1

### LINKED ISSUE #3843: Inf in BAND_*.dat and Missing 1 to 36 in `BAND_2.dat`
### Describe the bug

Fe, 2 atoms, nspin = 2, inf in `BAND_1.dat` and `BAND_2.dat`. Missing 1 to 36 in `BAND_2.dat`.
image
image

### Expected behavior

_No response_

### To Reproduce

[20240329-band-inf.zip](https://github.com/deepmodeling/abacus-develop/files/14799009/20240329-band-inf.zip)

### Environment

ABACUS v3.5.4

### Additional Context

_No response_

### Task list for Issue attackers (only for developers)

- [ ] Verify the issue is not a duplicate.
- [ ] Describe the bug.
- [ ] Steps to reproduce.
- [ ] Expected behavior.
- [ ] Error message.
- [ ] Environment details.
- [ ] Additional context.
- [ ] Assign a priority level (low, medium, high, urgent).
- [ ] Assign the issue to a team member.
- [ ] Label the issue with relevant tags.
- [ ] Identify possible related issues.
- [ ] Create a unit test or automated test to reproduce the bug (if applicable).
- [ ] Fix the bug.
- [ ] Test the fix.
- [ ] Update documentation (if necessary).
- [ ] Close the issue and inform the reporter (if applicable).

### PR DESCRIPTION
**Pull request by Qianruipku**
### Reminder
- [x] Have you linked an issue with this pull request?
- [x] Have you added adequate unit tests and/or case tests for your pull request?
- [x] Have you noticed possible changes of behavior below or in the linked issue?
- [x] Have you explained the changes of codes in core modules of ESolver, HSolver, ElecState, Hamilt, Operator or Psi? (ignore if not applicable)

### Linked Issue
Fix #3843

### Unit Tests and/or Case Tests for my changes
- A unit test is added for each new feature or bug fix.

### What's changed?
- And a function to gather k-point vectors.

### CODE REVIEW

File: source/module_cell/parallel_kpoints.cpp [hunk 1]
```diff
@@ -91,6 +91,26 @@ void Parallel_Kpoints::get_startk_pool(const int &nkstot)
     }
     return;
 }
+
+void Parallel_Kpoints::gatherkvec(const std::vector<ModuleBase::Vector3<double>>& vec_local,
+                                  std::vector<ModuleBase::Vector3<double>>& vec_global) const
+{
+    int nks = nks_pool[GlobalV::MY_POOL];
```
Comment (mohanchen) [inline source/module_cell/parallel_kpoints.cpp]: do not use GlobalV, set them as input parameters

Comment (mohanchen) [reply source/module_cell/parallel_kpoints.cpp]: int nks is a 'const' type?

File: source/module_cell/parallel_kpoints.cpp [hunk 2]
```diff
@@ -91,6 +91,26 @@ void Parallel_Kpoints::get_startk_pool(const int &nkstot)
...
+
+    int nks_tot = 0;
+    for (int i=0; i<GlobalV::KPAR; i++)
```
Comment (mohanchen) [inline source/module_cell/parallel_kpoints.cpp]: Do not use GlobalV::KPAR

File: source/module_cell/parallel_kpoints.cpp (hunk 1 above)
Comment (Qianruipku) [reply source/module_cell/parallel_kpoints.cpp]: MY_POOL is not read in from the INPUT file？

Comment (Qianruipku) [reply source/module_cell/parallel_kpoints.cpp]: OK

File: source/module_cell/parallel_kpoints.h [hunk 3]
```diff
@@ -21,15 +22,31 @@ class Parallel_Kpoints
...
+	// information about pool, dim: GlobalV::KPAR
 	int *nproc_pool;
 	int *startpro_pool;
 
-	// inforamation about kpoints //qianrui add comment
+	// inforamation about kpoints, dim: GlobalV::KPAR
 	int* nks_pool; //number of k-points in each pool
 	int* startk_pool; //the first k-point in each pool
-	int* whichpool; //whichpool[k] : the pool which k belongs to
+	int kpar; //number of pools
```
Comment (mohanchen) [inline source/module_cell/parallel_kpoints.h]: initialize with default values for variables: int kpar = 0

File: source/module_cell/parallel_kpoints.h [hunk 4]
```diff
@@ -21,15 +22,31 @@ class Parallel_Kpoints
...
+	
+	// information about which pool each k-point belongs to, 
+	int* whichpool; //whichpool[k] : the pool which k belongs to, dim: nkstot_np
```
Comment (mohanchen) [inline source/module_cell/parallel_kpoints.h]: Initialize pointers with nullptr, for example: int * whichpool = nullptr;

File: source/module_cell/parallel_kpoints.h (hunk 3 above)
Comment (Qianruipku) [reply source/module_cell/parallel_kpoints.h]: OK

File: source/module_cell/parallel_kpoints.h (hunk 4 above)
Comment (Qianruipku) [reply source/module_cell/parallel_kpoints.h]: OK

**Review by mohanchen — APPROVED** [review]

**Pull request merged**